extensions/panache/panache-hibernate-common/runtime/src/main/java/io/quarkus/panache/hibernate/common/runtime/PanacheJpaUtil.java

```diff
@@ -36,81 +29,6 @@ public class PanacheJpaUtil {
-            // this one cannot be null
-            String firstSelection = selectMatcher.group(1).trim();
-            String firstSelectionForMatching = firstSelection.toLowerCase(Locale.ROOT);
-            if (firstSelectionForMatching.startsWith("distinct")) {
-                // if firstSelection matched distinct only, we have something wrong in our selection list, probably functions/parens
-                // so bail out
-                if (firstSelectionForMatching.length() == 8) {
-                    return getCountQueryUsingParser(query);
-                }
-                // this one can be null
-                String secondSelection = selectMatcher.group(2);
-                // we can only count distinct single columns
-                if (secondSelection != null && !secondSelection.trim().isEmpty()) {
-                    throw new PanacheQueryException("Count query not supported for select query: " + query);
-                }
-                countQuery = "SELECT COUNT(" + firstSelection + ") " + selectMatcher.group(3);
-            } else {
-                // it's not distinct, forget the column list
-                countQuery = "SELECT COUNT(*) " + selectMatcher.group(3);
-            }
-        } else if (LONE_SELECT_PATTERN.matcher(query).matches()) {
-            // a select anywhere else in there might be tricky
-            return getCountQueryUsingParser(query);
-        } else if (FROM_PATTERN.matcher(query).matches()) {
-            countQuery = "SELECT COUNT(*) " + query;
-        } else {
-            throw new PanacheQueryException("Count query not supported for select query: " + query);
-        }
-
-        // remove the order by clause
-        String lcQuery = countQuery.toLowerCase();
-        int orderByIndex = lcQuery.lastIndexOf(" order by ");
-        if (orderByIndex != -1) {
-            countQuery = countQuery.substring(0, orderByIndex);
-        }
-        return countQuery;
-    }
-
-    /**
-     * This turns an HQL (already expanded from Panache-QL) query into a count query, using the
-     * ORM HQL parser. Slow version, see {@link #getFastCountQuery(String)} for the fast version.
-     */
-    public static String getCountQueryUsingParser(String query) {
-        HqlLexer lexer = new HqlLexer(CharStreams.fromString(query));
-        CommonTokenStream tokens = new CommonTokenStream(lexer);
-        HqlParser parser = new HqlParser(tokens);
-        SelectStatementContext statement = parser.selectStatement();
-        try {
-            CountParserVisitor visitor = new CountParserVisitor();
-            statement.accept(visitor);
-            return visitor.result();
-        } catch (RequiresSubqueryException x) {
-            // no luck
-            SubQueryAliasParserVisitor visitor = new SubQueryAliasParserVisitor();
-            statement.accept(visitor);
-            String ret = visitor.result();
-            return "select count( * ) from ( " + ret + " )";
-        }
-    }
-
```

Comment:
Definitely an improvement to get rid of this :fearful:

Reply:
Yeah, and now it's ORM's job to handle fast/slow cases :)